File: 07/prog.py
import re
import sys

def has_abba(s):
    for i in range(len(s) - 3):
        if s[i] == s[i+3] and s[i+1] == s[i+2] and s[i] != s[i+1]:
            return True
    return False

def has_aba_bab(outsides, insides):
    for text in outsides:
        for i in range(len(text) - 2):
            if text[i] == text[i+2] and text[i] != text[i+1]:
                if has_bab(insides, text[i], text[i+1]):
                    return True
    return False

def has_bab(insides, a, b):
    for text in insides:
        for i in range(len(text) - 2):
            if text[i] == b and text[i+1] == a and text[i+2] == b:
                return True
    return False

r = re.compile('[^\[\]]+')
count1 = 0
count2 = 0

while True:
    line = sys.stdin.readline().rstrip()
    if not line: break

    parts = r.findall(line)
    outsides = [parts[i] for i in range(0,len(parts),2)]
    insides  = [parts[i] for i in range(1,len(parts),2)]

    part1_match = False
    for text in outsides:
        if has_abba(text):
            part1_match = True
            break
    if part1_match:
        for text in insides:
            if has_abba(text):
                part1_match = False
                break
    if part1_match:
        count1 += 1

    if has_aba_bab(outsides, insides):
        count2 += 1

print(count1)
print(count2)

File: 07/prog.rs
use std::io;

fn check_abba(chars: &[char]) -> bool {
    for window in chars.windows(4) {
        if window[0] == window[3] && window[1] == window[2] && window[0] != window[1] {
            return true;
        }
    }
    false
}

fn check_aba_bab(outsides: &[Vec<char>], insides: &[Vec<char>]) -> bool {
    for chars in outsides {
        for window in chars.windows(3) {
            if window[0] == window[2] && window[0] != window[1] {
                if check_bab(insides, window[0], window[1]) {
                    return true;
                }
            }
        }
    }
    false
}

fn check_bab(insides: &[Vec<char>], a: char, b: char) -> bool {
    for chars in insides {
        for window in chars.windows(3) {
            if window[0] == b && window[1] == a && window[2] == b {
                return true;
            }
        }
    }
    false
}

fn main() {
    let mut count1 = 0;
    let mut count2 = 0;
    let mut line = String::new();
    while let Ok(n) = io::stdin().read_line(&mut line) {
        if n == 0 { break; }

        let mut in_brackets = false;
        let mut insides  = Vec::<Vec<char>>::new();
        let mut outsides = Vec::<Vec<char>>::new();
        let mut buf = Vec::<char>::new();
        for c in line.chars() {
            if c == '[' || c == '\n' {
                assert!(!in_brackets);
                in_brackets = true;
                outsides.push(std::mem::replace(&mut buf, vec![]));
            } else if c == ']' {
                assert!(in_brackets);
                in_brackets = false;
                insides.push(std::mem::replace(&mut buf, vec![]));
            } else {
                buf.push(c);
            }
        }

        let mut valid = false;
        for chars in &outsides {
            if check_abba(&chars) {
                valid = true;
                break;
            }
        }
        if valid {
            for chars in &insides {
                if check_abba(&chars) {
                    valid = false;
                    break;
                }
            }
        }
        if valid {
            count1 += 1;
        }

        if check_aba_bab(&outsides, &insides) {
            count2 += 1;
        }

        line.clear();
    }
    println!("{}", count1);
    println!("{}", count2);
}
